src/pkcs11.c

```diff
@@ -101,11 +101,60 @@ static struct {
 #endif
 };
 
+static CK_RV module_ensure(struct pkcs11_module *pkcs, struct dbg *dbg)
+{
+	CK_C_INITIALIZE_ARGS args = {
+		.flags = CKF_OS_LOCKING_OK,
+	};
+	CK_RV ck_rv;
+	int rv;
+
+	if (!pkcs || !dbg)
+		return CKR_ARGUMENTS_BAD;
+
+	/* check state unlocked */
+	if (pkcs->state == PKCS11_INITIALIZED)
+		return CKR_OK;
+
+	rv = pthread_mutex_lock(&pkcs->mutex);
+	if (rv) {
+		ps_dbg_error(dbg, "pkcs: %p, unable to lock module: %d", pkcs, rv);
+		return CKR_CANT_LOCK;
+	}
+
+	/* check state again under lock */
+	if (pkcs->state == PKCS11_INITIALIZED) {
+		ck_rv = CKR_OK;
+		goto out;
+	}
+
+	args.pReserved = (void *)pkcs->initargs;
```

Comment:
Maybe add a debug printout here so that we can see in the trace whenever a re-init happens. This would help to see whats going on in fork environments.

Reply:
ok, I'll address this in the next version.